accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,969 @@
+`V`).  Similarly for other member accesses.*
+
+Let _DV_ be an inline class declaration named `Inline` with type
+parameters
+<code>X<sub>1</sub> extends B<sub>1</sub>, .. X<sub>s</sub> extends B<sub>s</sub></code>.
+Assume that _DV_ declares a final instance variable with name `id` and
+type `R`.
+
+We say that the _declared representation type_ of `Inline`
+is `R`, and the _instantiated representation type_ corresponding to
+<code>Inline&lt;T<sub>1</sub>,.. T<sub>s</sub>&gt;</code> is
+<code>[T<sub>1</sub>/X<sub>1</sub>, .. T<sub>s</sub>/X<sub>s</sub>]R</code>.
+
+We will omit 'declared' and 'instantiated' from the phrase when it is
+clear from the context whether we are talking about the inline class
+itself, or we're talking a particular instantiation of a generic
+inline. *For non-generic inline classes, the representation type is
+the same in either case.*
+
+Let `V` be an inline type of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
+and let `R` be the corresponding instantiated representation type.
+`V` is a proper subtype of `Object?`. If `R` is non-nullable then `V`
+is a proper subtype of `Object` as well.
+
+*That is, an expression of an inline type can be assigned to a top type
+(like all other expressions), and if the representation type is
+non-nullable then it can also be assigned to `Object`. Non-inline types
+(except bottom types) cannot be assigned to inline types without a cast.*
+
+In the body of a member of an inline class declaration _DV_ named
+`Inline` and declaring the type parameters
+<code>X<sub>1</sub>, .. X<sub>s</sub></code>,
+the static type of `this` is
+<code>Inline&lt;X<sub>1</sub> .. X<sub>s</sub>&gt;</code>.
+The static type of the representation name is the representation
+type.
+
+*For example, in `inline class V { final R id; ...}`, `id` has type
+`R`, and `this` has type `V`.*
+
+Again, let `V` be an inline type of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
+and let `R` be the corresponding instantiated representation type.
+If `R` is not an inline type then we say that `V` is an inline type
+at level zero. If `R` is an inline type at level _k_ then we say that
+`V` is an inline type at level _k + 1_.
+A compile-time error occurs if the level of `V` is undefined.
+
+*In other words, cycles are not allowed.*
+
+For every inline class declaration _DV_ named `V`, it is a
+compile-time error if the inline type level of the raw type `V`
+is undefined.
+
+*In other words, we check for this kind of cycles on every non-generic
+inline class declaration using the type directly (even in the case
+where nobody uses that type), and we check for this kind of cycles on
+the instantiation-to-bound of every generic inline class declaration
+(again, even when nobody uses it). Finally, we check for this kind of
+cycles with every parameterized type `V<T1..Tk>` that actually
+occurs in the code which is being analyzed/compiled.*
+
```

Comment:
Yes! Alternatively, we could handle variance of the type arguments of an inline class in the same way as we handle variance for a `typedef` with type arguments: Deriving it from the occurrences of the type parameters in the body of the `typedef` (respectively: in the representation type of the inline class).

For now, I've added a compile-time error for the non-covariant occurrences, which allows us to do the `typedef` thing as an enhancement (and it will be possible to do the same thing with declaration-site variance, anyway).